tsfc/coffee_mode.py

```diff
@@ -0,0 +1,292 @@
+    return tuple(optimal_atomics)
+
+
+def factorise_atomics(monomial_sum, optimal_atomics, argument_indices):
+    """Group and factorise monomials using a list of atomics as common
+    subexpressions. Create new monomials for each group and optimise them recursively.
+
+    :arg monomial_sum: A :class:`MonomialSum` object, or a iterable collection
+                       of monomials, all of the monomials should have the
+                       same sum indices
+    :arg optimal_atomics: list of tuples of atomics to be used as common subexpression
+    :arg argument_indices: tuple of argument indices
+
+    :returns: a factorised :class:`MonomialSum` object, or the original object
+    if no changes are made. If original input is an iterable of :class`Monomial`s,
+    convert it to a :class:`MonomialSum` object.
+    """
+    #TODO: decide on input/output choice, iterable of monomials or MonomialSum
+    if not optimal_atomics or len(monomial_sum) < 2:
+        # Nothing to do
+        if isinstance(monomial_sum, MonomialSum):
+            return monomial_sum
+        else:
+            # input is list of monomials
+            new_monomial_sum = MonomialSum()
+            for m in monomial_sum:
+                new_monomial_sum.add(*m)
+            return new_monomial_sum
+
+    # Group monomials with respect to each optimal atomic
+    def group_key(monomial):
+        for oa in optimal_atomics:
+            if oa in monomial.atomics:
+                return oa
+        assert False, "Expect at least one optimal atomic per monomial."
+    factor_group = groupby(monomial_sum, key=group_key)
+
+    # We should not drop monomials
+    assert sum(len(ms) for _, ms in factor_group) == len(monomial_sum)
+
+    sum_indices = next(iter(monomial_sum)).sum_indices
+    new_monomial_sum = MonomialSum()
+    for oa, monomials in factor_group:
+        # Create new MonomialSum for the factorised out terms
+        sub_monomial_sum = MonomialSum()
+        for monomial in monomials:
+            atomics = list(monomial.atomics)
+            atomics.remove(oa)  # remove common factor
+            sub_monomial_sum.add((), atomics, monomial.rest)
+        sub_monomial_sum = optimise_monomials(sub_monomial_sum, argument_indices)
+        assert len(sub_monomial_sum) > 0
+        if len(sub_monomial_sum) == 1:
+            # result is a product, add back the common atomics then add to
+            # new MonomialSum directly
+            sub_monomial, = sub_monomial_sum
+            new_monomial_sum.add(sum_indices, sub_monomial.atomics + (oa,), sub_monomial.rest)
+        else:
+            # result is a sum, we need to create a new node
+            node = monomial_sum_to_expression(sub_monomial_sum)
+            if set(argument_indices) & set(node.free_indices):
+                new_monomial_sum.add(sum_indices, (oa, node), one)
+            else:
+                new_monomial_sum.add(sum_indices, (oa, ), node)
+    return new_monomial_sum
```

Comment:
I'm thinking, is this function returning a `MonomialSum` because this function is returning a `MonomialSum`? Or is there a deeper reason?